libsolidity/ast/Types.cpp

```diff
@@ -53,6 +53,78 @@ using namespace solidity::frontend;
 namespace
 {
 
+struct TypeComp
+{
+	bool operator()(Type const* lhs, Type const* rhs) const
+	{
+		solAssert(lhs && rhs, "");
+		return lhs->richIdentifier() < rhs->richIdentifier();
+	}
+};
+using TypeSet = std::set<Type const*, TypeComp>;
+
+bigint storageSizeUpperBound(
+	Type const& _type,
+	set<StructDefinition const*>& _structsSeen,
+	TypeSet& _oversizedSubtypes
+)
+{
+	switch (_type.category())
+	{
+	case Type::Category::Array:
+	{
+		auto const& t = dynamic_cast<ArrayType const&>(_type);
+		bigint baseTypeSize = storageSizeUpperBound(*t.baseType(), _structsSeen, _oversizedSubtypes);
+		if (baseTypeSize >= bigint(1) << 64)
+			_oversizedSubtypes.insert(t.baseType());
+		if (!t.isDynamicallySized())
+			return storageSizeUpperBound(*t.baseType(), _structsSeen, _oversizedSubtypes) * t.length();
+		break;
+	}
+	case Type::Category::Struct:
+	{
+		auto const& t = dynamic_cast<StructType const&>(_type);
+		bigint size = 1;
+		if (_structsSeen.count(&t.structDefinition()))
+			return size;
+		_structsSeen.insert(&t.structDefinition());
+		for (auto const& m: t.members(nullptr))
+			size += storageSizeUpperBound(*m.type, _structsSeen, _oversizedSubtypes);
+		_structsSeen.erase(&t.structDefinition());
+		return size;
+	}
+	case Type::Category::Mapping:
+	{
+		auto const* valueType = dynamic_cast<MappingType const&>(_type).valueType();
+		bigint valueTypeSize = storageSizeUpperBound(*valueType, _structsSeen, _oversizedSubtypes);
+		if (valueTypeSize >= bigint(1) << 64)
+			_oversizedSubtypes.insert(valueType);
+		break;
+	}
+	default:
+		break;
+	}
+	return bigint(1);
+}
+}
+
+namespace solidity::langutil
+{
+
+bigint storageSizeUpperBound(frontend::Type const& _type, std::vector<frontend::Type const*>* _oversizedSubtypes)
```

Comment:
Can be a virtual `Type::storageSizeUpperBound`, but it is also nice to keep the entire logic in one place.